test/testRules.js

```diff
@@ -1183,6 +1184,324 @@ export const testRules = {
+                locale: "de-DE",
+                source: " some source string ",
+                id: "resource-edge-whitespace.both-spaces-missing",
+                description: "Leading whitespace in target does not match leading whitespace in source",
+                highlight: `Source: "<e0> </e0>some…" Target: "<e1></e1>some…"`,
+            }),
+            new Result({
+                rule,
+                severity: "error",
+                pathName: "x/y",
+                locale: "de-DE",
+                source: " some source string ",
+                id: "resource-edge-whitespace.both-spaces-missing",
+                description: "Trailing whitespace in target does not match trailing whitespace in source",
+                highlight: `Source: "…ring<e0> </e0>" Target: "…ring<e1></e1>"`,
+            }),
+        ]);
+        test.done();
+    },
+
+    testResourceEdgeWhitespaceSpacesOnlyMatch: function(test) {
+        test.expect(2);
+
+        const rule = new ResourceEdgeWhitespace();
+        test.ok(rule);
+
+        const subject = {
+            // all-whitespace string
+            locale: "de-DE",
+            file: "x/y",
+            resource: new ResourceString({
+                key: "resource-edge-whitespace.spaces-only-match",
+                sourceLocale: "en-US",
+                source: " ",
+                targetLocale: "de-DE",
+                target: " ",
+                pathName: "resource-edge-whitespace-test.xliff",
+                state: "translated",
+            }),
+        };
+        const result = rule.match(subject);
+        test.equal(result, undefined); // for a valid resource match result should not be produced
+        test.done();
+    },
+
+    testResourceEdgeWhitespaceSpacesOnlyExtra: function(test) {
+        test.expect(2);
+
+        const rule = new ResourceEdgeWhitespace();
+        test.ok(rule);
+
+        const subject = {
+            // all-whitespace string
+            locale: "de-DE",
+            file: "x/y",
+            resource: new ResourceString({
+                key: "resource-edge-whitespace.spaces-only-extra",
+                sourceLocale: "en-US",
+                source: " ",
+                targetLocale: "de-DE",
+                target: "  ",
+                pathName: "resource-edge-whitespace-test.xliff",
+                state: "translated",
+            }),
+        };
+        const result = rule.match(subject);
+        test.deepEqual(
+            result,
+            new Result({
+                rule,
+                severity: "error",
+                pathName: "x/y",
+                locale: "de-DE",
+                source: " ",
+                id: "resource-edge-whitespace.spaces-only-extra",
+                description: "Leading whitespace in target does not match leading whitespace in source",
+                highlight: `Source: "<e0> </e0>" Target: "<e1>  </e1>"`,
+            })
+        );
         test.done();
     }
```

Comment:
What happens if the target string is empty or missing? Does this rule properly ignore it?

Reply:
This is accounted for in line 50: if either source or target is not a string (so also undefined), checks don't run on that resource

Reply:
What I meant is that we should probably have a unit test to verify those things. That way, if someone mucks with the function in the future, we have those guardrails. (We should probably look for code coverage tools for the ilib code.)

Reply:
Added in fcaa5fb

As for code coverage - maybe we should consider changing our testing libraries to something newer? Nodeunit has been deprecated 5 years ago 😄 Jest has builtin coverage support